statsmodels/regression/_tools.py

```diff
@@ -22,24 +19,34 @@ class _MinimalWLS(object):
         1d array of weights.  If you supply 1/W then the variables are pre-
         multiplied by 1/sqrt(W).  If no weights are supplied the default value
         is 1 and WLS reults are the same as OLS.
+    check_finite : bool, optional
+        Flag indicating whether to check for inf/nan in endog and weights.
+        If True and any are found, ValueError is raised.
 
     Notes
     -----
-    Need resid, scale, fittedvalues, model.weights!
+    Need resid, scale, fittedvalues, model.weights
         history['scale'].append(tmp_results.scale)
         if conv == 'dev':
             history['deviance'].append(self.deviance(tmp_results))
         elif conv == 'sresid':
             history['sresid'].append(tmp_results.resid/tmp_results.scale)
         elif conv == 'weights':
             history['weights'].append(tmp_results.model.weights)
-    Does not perform and checks on the input data
+    Does not perform any checks on the input data
     """
 
-    def __init__(self, endog, exog, weights=1.0):
+    def __init__(self, endog, exog, weights=1.0, check_finite=False):
         self.endog = endog
         self.exog = exog
         self.weights = weights
+        if check_finite:
+            msg = 'NaN/inf detected in {0}, estimation infeasible.'
+            if not np.all(np.isfinite(endog)):
+                raise ValueError(msg.format('endog'))
+            if not np.all(np.isfinite(weights)):
+                raise ValueError(msg.format('weights'))
+
```

Comment:
I think to catch algorithmic problems, e.g. negative weights, we need to do the check on w_half

Reply:
Is there a scenario where `w_half` is not OK but `weights` is? Computing `w_half` when there is an issue will likely raise a warning before the exception is thrown.

Reply:
The problems is negative weights which cause nans in w_half, which will cause a Warning but doesn't prevent the nans to go into linalg.

I don't remember or know in which part the problem was in my example when I ran into this, the reason for the problem was "user error". 
IIRC, we should get it in a linear probability model for GLM Binomial when predicted is outside open interval (0, 1), but maybe we "sanitize" it in this case by clipping.
There is a set of issues related to this #2208 where we intentionally don't have a properly defined model.

Reply:
I switched it around.